### PR TITLE
[WIP]: Add dcrt Lasso + Random Forest

### PR DESCRIPTION
**Pull request by achamma723**

### CODE REVIEW

File: examples/example_dcrt.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    # Random choice of the coefficients for the signal
+    effectset = np.array([-0.5, -1, -2, -3, 0.5, 1, 2, 3])
+    betas = np.random.choice(effectset, n_signal)
+
+    prod_signal = np.dot(X[:, :n_signal], betas)
+    sigma_noise = np.linalg.norm(prod_signal) / (snr * np.sqrt(X.shape[0]))
+
+    # Generation of the outcome
+    y = prod_signal + sigma_noise * np.random.normal(0, 1, size=X.shape[0])
+    y = np.array([max(0.0, i) for i in y])
```
Comment (bthirion) [inline examples/example_dcrt.py:38]: Why do you do that ?

File: examples/example_dcrt.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    np.random.seed(sim_ind)
+    DEBUG = False
+
+    n = 1000 if not DEBUG else 10
+    p = 10 if not DEBUG else 5
+    n_signal = 2 if not DEBUG else 1
+    snr = 4
+    rho = 0.8
+
+    # Create Correlation matrix with the toeplitz design
```
Comment (bthirion) [inline examples/example_dcrt.py]: This is not a Toeplitz matrix but a 1-block structure. To create a Toeplitz matrix, you need to specify diagonals.

File: examples/example_dcrt.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+dcrtLasso_type1 = []
+dcrtLasso_power = []
+dcrtForest_type1 = []
+dcrtForest_power = []
+
+for sim_ind in range(100):
+    print(f"Processing: {sim_ind+1}")
+    np.random.seed(sim_ind)
+    DEBUG = False
```
Comment (bthirion) [inline examples/example_dcrt.py]: To be removed

File: examples/example_dcrt.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    prod_signal = np.dot(X[:, :n_signal], betas)
+    sigma_noise = np.linalg.norm(prod_signal) / (snr * np.sqrt(X.shape[0]))
+
+    # Generation of the outcome
+    y = prod_signal + sigma_noise * np.random.normal(0, 1, size=X.shape[0])
+    y = np.array([max(0.0, i) for i in y])
+
+    ## dcrt Lasso ##
+    res_lasso = dcrt_zero(X, y, screening=False, verbose=True)
+    dcrtLasso_type1.append(sum(res_lasso[1][n_signal:] < 5e-2) / (p-n_signal))
```
Comment (bthirion) [inline examples/example_dcrt.py]: 5.e-2 should be a parameter, e.g. `alpha`, with an explicit meaning.

Comment (bthirion) [inline examples/example_dcrt.py]:
```suggestion
    dcrtLasso_type1.append(sum(res_lasso[1][n_signal:] < 5e-2) / (p - n_signal))
```

File: examples/example_dcrt.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    prod_signal = np.dot(X[:, :n_signal], betas)
+    sigma_noise = np.linalg.norm(prod_signal) / (snr * np.sqrt(X.shape[0]))
+
+    # Generation of the outcome
+    y = prod_signal + sigma_noise * np.random.normal(0, 1, size=X.shape[0])
+    y = np.array([max(0.0, i) for i in y])
+
+    ## dcrt Lasso ##
+    res_lasso = dcrt_zero(X, y, screening=False, verbose=True)
```
Comment (bthirion) [inline examples/example_dcrt.py]: `res_lasso` is ambiguous. 'res' could be 'results' or 'residuals' etc.

File: hidimstat/dcrt.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
+"""Implementation of distillation Conditional Randomization Test, by Liu et
+al. (2020) <https://arxiv.org/abs/2006.03980>. Currently only d0_CRT
+resampling-free is implemented.
```
Comment (bthirion) [inline hidimstat/dcrt.py]:
```suggestion
Currently, only resampling-free d0_CRT is implemented.
```

File: hidimstat/dcrt.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+
+def dcrt_zero(X, y, fdr=0.1, estimated_coef=None, Sigma_X=None, cv=5,
+              n_regus=20, max_iter=1000, use_cv=False, refit=False,
+              loss='least_square', screening=True, screening_threshold=1e-1,
+              scaled_statistics=False, statistic='residual', centered=True,
+              alpha=None, solver='liblinear', fdr_control='bhq', n_jobs=1,
+              verbose=False, joblib_verbose=0, ntree=100,
+              type_prob='regression', random_state=2022):
+
```
Comment (bthirion) [inline hidimstat/dcrt.py:23]: Missing docstring

Comment (bthirion) [reply hidimstat/dcrt.py:23]: Some parameters are unclear, e.g. `type_prob`, `n_regus`
Try to coerce to sklearn terminology whenever possible.

File: hidimstat/dcrt.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                    X, y, idx, Sigma_X=Sigma_X, cv=3, use_cv=use_cv,
+                    alpha=alpha, n_jobs=1, n_regus=n_regus,
+                    ntree=ntree, loss=loss, type_prob=type_prob,
+                    random_state=random_state)
+                for idx in selection_set)
+        else:
+            raise ValueError(f'{loss} loss is not supported.')
+    else:
+        raise ValueError(f'{statistic} statistic is not supported.')
+    Ts = np.zeros(n_features)
```
Comment (bthirion) [inline hidimstat/dcrt.py]: why not `ts` ?

File: hidimstat/dcrt.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        pvals = np.minimum(2 * stats.norm.sf(np.abs(Ts)), 1)
+    elif statistic == 'likelihood':
+        pvals = stats.chi2.sf(Ts, 1)
+
+    threshold = fdr_threshold(pvals, fdr=fdr, method=fdr_control)
+    selected = np.where(pvals <= threshold)[0]
+
+    if verbose:
+        return selected, pvals, Ts
+
```
Comment (bthirion) [inline hidimstat/dcrt.py:107]: Please leave a few comments on the code to help understanding

File: hidimstat/dcrt.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    return selected
+
+
+def _x_distillation_lasso(X, idx, Sigma_X=None, cv=3, n_regus=100, alpha=None,
+                          use_cv=False, n_jobs=1):
+
```
Comment (bthirion) [inline hidimstat/dcrt.py:113]: Adda one-line docstring to give an idea of what this is doing.

File: hidimstat/dcrt.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    n_samples = X.shape[0]
+    X_minus_idx = np.delete(np.copy(X), idx, 1)
+
+    if Sigma_X is None:
```
Comment (bthirion) [inline hidimstat/dcrt.py:117]: Explain a bit what you're doing

File: hidimstat/dcrt.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        clf_null.fit(X_minus_idx, y)
+        coef_minus_idx = clf_null.coef_
+    else:
+        coef_minus_idx = np.delete(np.copy(coef_full), idx)
+
+    eps_res = y - X_minus_idx.dot(coef_minus_idx)
+    sigma2_y = np.mean(eps_res ** 2)
+
+    # T follows Gaussian distribution
+    Ts = np.dot(eps_res, X_res) / np.sqrt(n_samples * sigma2_X * sigma2_y)
```
Comment (bthirion) [inline hidimstat/dcrt.py]:
```suggestion
    ts = np.dot(eps_res, X_res) / np.sqrt(n_samples * sigma2_X * sigma2_y)
```

File: hidimstat/dcrt.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+    """
+    X_minus_idx = np.delete(np.copy(X), 0, 1)
+
+    if loss == 'least_square':
+        clf_null = LassoCV(cv=cv, n_jobs=n_jobs, n_alphas=n_regus,
+                           random_state=0, max_iter=max_iter, tol=tol)
+        clf_null.fit(X_minus_idx, y)
+        optimal_lambda = clf_null.alpha_
+
+    return optimal_lambda
```
Comment (bthirion) [inline hidimstat/dcrt.py:201]: lambda or alpha ?

**Review by bthirion — COMMENTED** [review]
Remember that the code is meant to be read, not only executed ...

File: examples/example_dcrt.py (hunk 1 above)
Comment (achamma723) [reply examples/example_dcrt.py:38]: I'm adding gaussian noise to the computed signal before applying a relu function for the non-linearity

File: examples/example_dcrt.py (hunk 2 above)
Comment (achamma723) [reply examples/example_dcrt.py]: I've changed the design to target the standard design of the correlation matrix but I've missed correcting the comment

File: examples/example_dcrt.py (hunk 1 above)
Comment (bthirion) [reply examples/example_dcrt.py:38]: Explain what you do, and why !

File: examples/example_dcrt.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    p = 10
+    # Number of relevant variables
+    n_signal = 2
+    # Signal-to-noise ratio
+    snr = 4
+    # Correlation coefficient
+    rho = 0.8
+    # Nominal false positive rate
+    alpha = 5e-2
+
```
Comment (bthirion) [inline examples/example_dcrt.py:32]: Why don't you use  `multivariate_1D_simulation` from hidimstat.scenario ?

**Pull request closed without merging**